localization/ndt_scan_matcher/include/ndt_scan_matcher/map_update_module.hpp

```diff
@@ -44,36 +45,46 @@ class MapUpdateModule
   using PointTarget = pcl::PointXYZ;
   using NormalDistributionsTransform =
     pclomp::MultiGridNormalDistributionsTransform<PointSource, PointTarget>;
+  using NdtPtrType = std::shared_ptr<NormalDistributionsTransform>;
 
 public:
   MapUpdateModule(
     rclcpp::Node * node, std::mutex * ndt_ptr_mutex,
-    std::shared_ptr<NormalDistributionsTransform> ndt_ptr,
+    std::shared_ptr<NormalDistributionsTransform> & ndt_ptr,
     HyperParameters::DynamicMapLoading param);
 
 private:
   friend class NDTScanMatcher;
 
+  // Update the specified NDT
   void update_ndt(
+    NdtPtrType & ndt,
     const std::vector<autoware_map_msgs::msg::PointCloudMapCellWithID> & maps_to_add,
     const std::vector<std::string> & map_ids_to_remove);
   void update_map(const geometry_msgs::msg::Point & position);
-  [[nodiscard]] bool should_update_map(const geometry_msgs::msg::Point & position) const;
+  [[nodiscard]] bool should_update_map(const geometry_msgs::msg::Point & position);
   void publish_partial_pcd_map();
 
+  // Pre-fetch NDT map in a child thread
+  void prefetch_map(const geometry_msgs::msg::Point & position, NdtPtrType & ndt);
+
   rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr loaded_pcd_pub_;
 
   rclcpp::Client<autoware_map_msgs::srv::GetDifferentialPointCloudMap>::SharedPtr
     pcd_loader_client_;
 
-  std::shared_ptr<NormalDistributionsTransform> ndt_ptr_;
+  NdtPtrType & ndt_ptr_;
   std::mutex * ndt_ptr_mutex_;
   rclcpp::Logger logger_;
   rclcpp::Clock::SharedPtr clock_;
 
   std::optional<geometry_msgs::msg::Point> last_update_position_ = std::nullopt;
 
   HyperParameters::DynamicMapLoading param_;
+
+  // Indicate if there is a prefetch thread waiting for being collected
+  NdtPtrType secondary_ndt_ptr_;
+  bool rebuild_;
```

Comment:
Is it correct to think that this variable called `rebuild_` means "whether a rebuild is required?" In that case, how about naming it `need_rebuild_`?

Reply:
@SakodaShintaro Yes.
In this PR, I load the map to a secondary ndt ptr, so the scan alignment can be done on the main ndt ptr without locking. When the map loading is done, the main ndt ptr is swapped with the secondary ndt ptr, so the time of blocking is negligible (in my assumption).

However, when the new position is moved to a new position very far from the previous position, aligning scan on the main ndt ptr (while updating the secondary ndt) is not correct. In such cases, the main ndt ptr should be locked and updated _before_ any scan alignment.

This _rebuild__ flag is used to indicate whether the update should be done on the main ndt ptr directly. If it is false, the update is done on the secondary ndt ptr. If it is true, the update is performed directly on the ndt ptr and the pointer is locked until the update is finished.